File: packages/redact-node/test/redact.test.mjs
// The redact-node test suite. Runs through the WebAssembly runtime with model
// files loaded from the local LiteRT resources instead of the Hugging Face Hub.
import assert from "node:assert/strict";
import { test } from "node:test";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { Redact } from "../index.js";

const here = path.dirname(fileURLToPath(import.meta.url));
const directory = path.join(here, "../../../Sources/RedactTFLiteResources/Resources");

let redact;
let loadError;
try {
  redact = await Redact.load({ directory });
} catch (e) {
  loadError = e;
}
const modelOpts = redact ? {} : { skip: `model unavailable: ${String(loadError).slice(0, 80)}` };

test("redaction masks names, email, IBAN", modelOpts, async () => {
  const r = await redact.redaction("Email Anna Kovács at anna@example.hu, IBAN DE89370400440532013000.");
  assert.match(r.redactedText, /\[GIVEN_NAME_1\]/);
  assert.match(r.redactedText, /\[EMAIL_1\]/);
  assert.match(r.redactedText, /\[BANK_ACCOUNT_1\]/);
  assert.equal(r.items.find((i) => i.label === "EMAIL")?.original, "anna@example.hu");
});

test("addresses, VAT, IMEI redacted", modelOpts, async () => {
  const r = await redact.redaction("Ship to 123 Main Street, Apt 4B. VAT DE129273398, IMEI 490154203237518.");
  const got = new Set(r.items.map((i) => i.label));
  for (const l of ["BUILDING_NUMBER", "STREET_NAME", "SECONDARY_ADDRESS", "TAX_ID", "IMEI"]) {
    assert.ok(got.has(l), `expected ${l}`);
  }
});

test("restore round-trips exactly", modelOpts, async () => {
  const text = "Call Dr. Alice Grant on +49 30 1234567.";
  const r = await redact.redaction(text);
  assert.equal(r.restore(r.redactedText), text);
});

test("label filter", modelOpts, async () => {
  const r = await redact.redaction("Anna at anna@x.com, IBAN DE89370400440532013000.", { labels: ["EMAIL"] });
  assert.deepEqual(new Set(r.items.map((i) => i.label)), new Set(["EMAIL"]));
});

File: packages/redact-kotlin/src/androidTest/kotlin/ai/desertant/redact/RedactTest.kt
package ai.desertant.redact

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import kotlinx.coroutines.test.runTest
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Instrumented tests for the Android binding, exercising the real on-device path
 * via JNI: java.util.regex, platform JSON via CHostBridge, system ICU for NFKC,
 * and the static-stdlib runtime. The bundled model comes from the
 * `redact-tflite-resources` androidTest dependency.
 */
@RunWith(AndroidJUnit4::class)
class RedactTest {
    private lateinit var redact: Redact

    @Before fun setUp() { redact = Redact.bundled() }
    @After fun tearDown() { redact.close() }

    @Test fun redactionEndToEnd() = runTest {
        val r = redact.redaction("Email Anna Kovács at anna@example.hu, IBAN DE89370400440532013000.")
        assertTrue(r.redactedText, Regex("""\[GIVEN_NAME_1\]""").containsMatchIn(r.redactedText))
        assertTrue(Regex("""\[EMAIL_1\]""").containsMatchIn(r.redactedText))
        assertTrue(Regex("""\[BANK_ACCOUNT_1\]""").containsMatchIn(r.redactedText))
        assertEquals("anna@example.hu", r.items.first { it.label == "EMAIL" }.original)
    }

    @Test fun addressesVatImei() = runTest {
        val r = redact.redaction("Ship to 123 Main Street, Apt 4B. VAT DE129273398, IMEI 490154203237518.")
        val got = r.items.map { it.label }.toSet()
        for (l in listOf("BUILDING_NUMBER", "STREET_NAME", "SECONDARY_ADDRESS", "TAX_ID", "IMEI"))
            assertTrue("expected $l in $got", l in got)
    }

    @Test fun restoreRoundTrips() = runTest {
        val text = "Call Dr. Alice Grant on +49 30 1234567."
        val r = redact.redaction(text)
        assertEquals(text, r.restore(r.redactedText))
    }

    @Test fun labelFilter() = runTest {
        val r = redact.redaction("Call +34 600 100 200 or email me@x.com", Options(labels = setOf("PHONE")))
        assertTrue(r.items.all { it.label == "PHONE" })
        assertTrue(r.redactedText.contains("[PHONE_1]"))
        assertTrue(r.redactedText.contains("me@x.com"))
    }

    // On-demand download from the Hub via the Swift ModelStore (host HTTP +
    // POSIX cache), then redact. Network; the model caches under the app cacheDir.
    @Test fun downloadFromHubAndRedact() = runTest {
        val ctx = InstrumentationRegistry.getInstrumentation().targetContext
        val downloaded = Redact(ctx)
        try {
            downloaded.download()
            assertTrue(downloaded.isDownloaded())
            val r = downloaded.redaction("Email Anna Kovács at anna@example.com about the invoice.")
            assertTrue(r.redactedText, Regex("""\[EMAIL_1\]""").containsMatchIn(r.redactedText))
            assertTrue(r.redactedText, Regex("""\[GIVEN_NAME_1\]""").containsMatchIn(r.redactedText))
        } finally {
            downloaded.close()
        }
    }
}
